### ISSUE TITLE
Question/feature request: specific custom index SQL code

### ISSUE THREAD

**Issue by hi2u** [typeorm/typeorm#2561]
From my understanding, at present only basic indices on single or multiple columns can be created.

e.g. Looking at https://github.com/typeorm/typeorm/blob/master/docs/indices.md#disabling-synchronization I see that for indices that aren't supported, the suggestion is to create them using another migration system (or without one at all) and then tell TypeORM to ignore them.

Obviously all the different types of indices across all the databases would be very hard for TypeORM to have a syntax for, especially where conditional logic is thrown in too, and I don't think it's worth trying to make TypeORM aware of this complexitity as it's not really needed for ORM functionality, but is needed for a migration/diff/sync system.  So is it possible to instead just specify the raw SQL code used to create the index?

Using the example at the bottom of the page linked above, I imagine something like:

```
@CustomIndex("POST_NAME_INDEX", 'CREATE INDEX "POST_NAME_INDEX" ON "post" (lower("name")')
```
...or maybe even without the CREATE prefix and table name included seeing they might be redundant, so just...

```
@CustomIndex("POST_NAME_INDEX", '(lower("name")')
```

i.e. The 2nd argument is just a string of plain SQL code that is passed directly to the database.  TypeORM doesn't need to parse it.

This would allow anyone to create any type of index without having to implement a 2nd migration system to handle this stuff.  Having conditionals on NULLs etc is a pretty common thing to do in postgres, and a limitation that I've run into on the first table I've created with TypeORM.

To account for modifications to custom indices after they've been initially created, a couple of options could be:

* TypeORM just stays ignorant of them.  The user can just rename the index in order for it to be dropped and re-created (which is what gets done in SQL every time an index changes anyway).  So it makes no differences in terms of migration speed.
* TypeORM appends a checksum of the raw code to the index name and detects if it needs to be re-created from that.

On top of indices, something like this could also be useful to let us add views, triggers and functions etc into our TypeORM models so that we be use TypeORM sync + migrations for everything without needing a separate migration system.

I'm only just getting started with TypeORM, so please let me know if I've missed something, or there's another way to do this kind of stuff without using a 2nd migration system.

**Comment by pleerock** [comment]
Yes, we can do it. We already have such functionality for `@Check` decorator.

**Comment by nickfla1** [comment]
Any update on this?

**Comment by nickfla1** [comment]
Sorry for the double post but do you accept PRs for this issue?

**Comment by brianbegy** [comment]
Any update?  We need this too.

**Comment by gabbrieu** [comment]
any update?
